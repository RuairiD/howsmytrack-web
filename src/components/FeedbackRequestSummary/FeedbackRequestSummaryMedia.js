import React from "react";

import { Alert } from "antd";
import MediaEmbed from "../MediaEmbed/MediaEmbed";

const FeedbackRequestSummaryMedia = ({ mediaUrl, mediaType }) => {
    if (!mediaUrl) {
        return (
            <Alert
                message="No track has been provided for this request. You'll be able to join a group and provide feedback for others, but you won't receive any feedback for yourself."
                type="info"
                showIcon
            />
        );
    }
    if (!mediaType) {
        return (
            <Alert
                message="Unable to play media. Invalid URL."
                type="error"
                showIcon
            />
        );
    }
    return <MediaEmbed mediaUrl={mediaUrl} mediaType={mediaType} />;
};

export default FeedbackRequestSummaryMedia;
