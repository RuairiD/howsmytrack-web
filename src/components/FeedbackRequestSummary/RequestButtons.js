import React from "react";

import { Button, Popconfirm } from "antd";

const getDeleteConfirmationText = (mediaUrl) => {
    if (mediaUrl) {
        return "This request has not been assigned to a group. If you delete it, you will not receive any feedback on it. Are you sure you want to delete this request?";
    }
    return "This request has not been assigned to a group. Are you sure you want to delete this request?";
};

const RequestButtons = ({ mediaUrl, onEditClick, onDelete }) => (
    <React.Fragment>
        <Button
            shape="circle"
            icon="edit"
            onClick={onEditClick}
            style={{
                marginRight: "0.5em",
            }}
            className="request-button"
        />
        <Popconfirm
            title={getDeleteConfirmationText(mediaUrl)}
            onConfirm={onDelete}
            okText="Yes"
            cancelText="No"
        >
            <Button
                shape="circle"
                icon="delete"
                className="request-button"
            />
        </Popconfirm>
    </React.Fragment>
);

export default RequestButtons;
