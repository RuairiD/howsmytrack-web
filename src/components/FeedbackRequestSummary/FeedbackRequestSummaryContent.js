import React from "react";

import { Row, Col, Typography } from "antd";
import { Div } from "lemon-reset";
import GENRE_OPTIONS from "../../genres";
import type { FeedbackRequestSummaryProps } from "./FeedbackRequestSummary";
import FeedbackRequestSummaryMedia from "./FeedbackRequestSummaryMedia";

type Props = {
    feedbackRequestSummary: FeedbackRequestSummaryProps,
};

/*
 * Component for inline feedback request preview shown on FeedbackGroupsPage and in FeedbackRequestForm
 */
const FeedbackRequestSummaryContent = ({ feedbackRequestSummary }: Props) => (
    <Div>
        <Row gutter={[16, 16]}>
            <Col>
                <FeedbackRequestSummaryMedia
                    mediaUrl={feedbackRequestSummary.mediaUrl}
                    mediaType={feedbackRequestSummary.mediaType}
                />
            </Col>
        </Row>
        <Row gutter={[16, 16]}>
            <Col>
                <Typography.Text strong>Genre: </Typography.Text>
                <Typography.Text>{GENRE_OPTIONS[feedbackRequestSummary.genre]}</Typography.Text>
            </Col>
        </Row>
        {feedbackRequestSummary.mediaUrl && (
            <Row>
                {feedbackRequestSummary.feedbackPrompt && (
                    <Col>
                        <Typography.Text strong>You said: </Typography.Text>
                        <Typography.Text>"{feedbackRequestSummary.feedbackPrompt}"</Typography.Text>
                    </Col>
                )}
                {!feedbackRequestSummary.feedbackPrompt && (
                    <Col>
                        <Typography.Text strong>You did not provide any additional information for feedback.</Typography.Text>
                    </Col>
                )}
            </Row>
        )}
    </Div>
);

export default FeedbackRequestSummaryContent;
