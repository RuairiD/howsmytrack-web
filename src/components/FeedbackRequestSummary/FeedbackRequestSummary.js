import React from 'react';

import apiRoot from '../../apiRoot';

import { Button, Card, message, Popconfirm, Result, Spin } from 'antd';
import EditFeedbackRequestModal from '../EditFeedbackRequestModal/EditFeedbackRequestModal';
import FeedbackRequestSummaryContent from './FeedbackRequestSummaryContent';

export type FeedbackRequestSummaryProps = {
    feedbackRequestId: number,
    mediaUrl: string,
    mediaType: string,
    feedbackPrompt: string,
    emailWhenGrouped: boolean,
};

type Props = {
    feedbackRequestSummary: FeedbackRequestSummaryProps,
    showButtons: boolean,
};

type State = {
    isEditFeedbackRequestModalVisible: boolean,
    deleteRequestSent: boolean,
    requestDeleted: boolean,
};

const DELETE_FEEDBACK_REQUEST_MUTATION = `mutation DeleteFeedbackRequest($feedbackRequestId: Int!) {
    deleteFeedbackRequest(feedbackRequestId: $feedbackRequestId) {
        success
        error
    }
}`;

class FeedbackRequestSummary extends React.Component<Props, State> {
    /*
     * Component for inline feedback request preview shown on FeedbackGroupsPage
     * which also allows users to edit their requests.
     */
    state = {
        isEditFeedbackRequestModalVisible: false,
        deleteRequestSent: false,
        requestDeleted: false,
    };

    showEditFeedbackRequestModal = () => {
        this.setState({
            isEditFeedbackRequestModalVisible: true,
        })
    };

    hideEditFeedbackRequestModal = () => {
        this.setState({
            isEditFeedbackRequestModalVisible: false,
        })
    };

    deleteRequest = () => {
        this.setState({
            deleteRequestSent: true,
        })
        fetch(apiRoot +'/graphql/', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
            body: JSON.stringify({
                query: DELETE_FEEDBACK_REQUEST_MUTATION,
                variables: {
                    feedbackRequestId: this.props.feedbackRequestSummary.feedbackRequestId,
                },
            }),
            credentials: 'include',
        }).then(result =>
            result.json()
        ).then((data) => {
            if (!data['data']['deleteFeedbackRequest'].success) {
                message.error(data['data']['deleteFeedbackRequest'].error);
            }
            this.setState({
                requestSent: false,
                requestDeleted: data['data']['deleteFeedbackRequest'].success,
            });
        });
    };

    renderButtons = () => {
        return (
            <React.Fragment>
                <Button
                    shape="circle"
                    icon="edit"
                    onClick={this.showEditFeedbackRequestModal}
                    style={{
                        marginRight: '0.5em',
                    }}
                />
                <Popconfirm
                    title="This request has not been assigned to a group. If you delete it, you will not receive any feedback on it. Are you sure you want to delete this request?"
                    onConfirm={this.deleteRequest}
                    okText="Yes"
                    cancelText="No"
                >
                    <Button
                        shape="circle"
                        icon="delete"
                    />
                </Popconfirm>
            </React.Fragment>
        )
    }

    render() {
        if (this.state.requestDeleted) {
            return (
                <Result
                    status="success"
                    title="This request has been deleted."
                />
            )
        }
        return (
            <Spin spinning={this.state.deleteRequestSent}>
                <Card
                    title="You submitted:"
                    extra={this.props.showButtons && this.renderButtons()}
                >
                    <FeedbackRequestSummaryContent
                        feedbackRequestSummary={this.props.feedbackRequestSummary}
                    />
                </Card>
                <EditFeedbackRequestModal
                    onCancel={this.hideEditFeedbackRequestModal}
                    isVisible={this.state.isEditFeedbackRequestModalVisible}
                    feedbackRequestId={this.props.feedbackRequestSummary.feedbackRequestId}
                    mediaUrl={this.props.feedbackRequestSummary.mediaUrl}
                    feedbackPrompt={this.props.feedbackRequestSummary.feedbackPrompt}
                    emailWhenGrouped={this.props.feedbackRequestSummary.emailWhenGrouped}
                />
            </Spin>
        );
    }
}

export default FeedbackRequestSummary;
