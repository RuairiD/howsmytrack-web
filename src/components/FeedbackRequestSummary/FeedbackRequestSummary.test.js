import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import waitForExpect from "wait-for-expect";
import { act } from "react-dom/test-utils";
import axios from "axios";
import { message } from "antd";
import FeedbackRequestSummary from "./FeedbackRequestSummary";

jest.mock("react-ga");
jest.mock("axios");

const errorMessageSpy = jest.spyOn(message, "error");

describe("FeedbackRequestSummary", () => {
    beforeEach(() => {
        axios.post.mockRestore();
    });

    it("renders a summary of an existing feedback request, including media and buttons to edit and delete the request", () => {
        const wrapper = shallow(
            <FeedbackRequestSummary
                feedbackRequestSummary={{
                    feedbackRequestId: 1901,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: true,
                    genre: "genre",
                }}
                showButtons
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it("renders a summary of an existing feedback request without buttons if showButtons is omitted", () => {
        const wrapper = shallow(
            <FeedbackRequestSummary
                feedbackRequestSummary={{
                    feedbackRequestId: 1901,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: true,
                    genre: "genre",
                }}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("shows an EditFeedbackRequestModal when the edit button is clicked", () => {
        const wrapper = mount(
            <FeedbackRequestSummary
                feedbackRequestSummary={{
                    feedbackRequestId: 1901,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: true,
                    genre: "genre",
                }}
                showButtons
            />,
        );

        act(() => wrapper.find("RequestButtons").get(0).props.onEditClick());

        wrapper.update();
        expect(wrapper.find("EditFeedbackRequestModal").get(0).props.isVisible).toBeTruthy();
    });

    it("shows the EditFeedbackRequestModal when dismissed", () => {
        const wrapper = mount(
            <FeedbackRequestSummary
                feedbackRequestSummary={{
                    feedbackRequestId: 1901,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: true,
                    genre: "genre",
                }}
                showButtons
            />,
        );

        act(() => wrapper.find("RequestButtons").get(0).props.onEditClick());
        act(() => wrapper.find("EditFeedbackRequestModal").get(0).props.onCancel());

        wrapper.update();
        expect(wrapper.find("EditFeedbackRequestModal").get(0).props.isVisible).not.toBeTruthy();
    });

    it("makes an API request to delete the feedback request when delete button is clicked, and shows a confirmation", async () => {
        axios.post.mockResolvedValueOnce({
            data: {
                data: {
                    deleteFeedbackRequest: {
                        success: true,
                        error: null,
                    },
                },
            },
        });

        const wrapper = mount(
            <FeedbackRequestSummary
                feedbackRequestSummary={{
                    feedbackRequestId: 1901,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: true,
                    genre: "genre",
                }}
                showButtons
            />,
        );

        expect(wrapper.find("Result").length).toBe(0);
        expect(wrapper.find("Result[status=\"success\"]").length).toBe(0);

        await act(async () => wrapper.find("RequestButtons").get(0).props.onDelete());

        await waitForExpect(() => {
            wrapper.update();
            expect(wrapper.find("Result").length).toBe(1);
            expect(wrapper.find("Result[status=\"success\"]").length).toBe(1);
        });
    });

    it("shows an error if request deletion fails", async () => {
        axios.post.mockResolvedValueOnce({
            data: {
                data: {
                    deleteFeedbackRequest: {
                        success: false,
                        error: "error",
                    },
                },
            },
        });

        const wrapper = mount(
            <FeedbackRequestSummary
                feedbackRequestSummary={{
                    feedbackRequestId: 1901,
                    mediaUrl: "mediaUrl",
                    mediaType: "mediaType",
                    feedbackPrompt: "feedbackPrompt",
                    emailWhenGrouped: true,
                    genre: "genre",
                }}
                showButtons
            />,
        );

        await act(async () => wrapper.find("RequestButtons").get(0).props.onDelete());

        await waitForExpect(() => {
            wrapper.update();
            expect(errorMessageSpy).toHaveBeenCalled();
        });
    });
});
