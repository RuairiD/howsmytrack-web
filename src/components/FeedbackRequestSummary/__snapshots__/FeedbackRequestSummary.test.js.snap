// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackRequestSummary renders a summary of an existing feedback request without buttons if showButtons is omitted 1`] = `
<Spin
  size="default"
  spinning={false}
  wrapperClassName=""
>
  <Card
    title="You submitted:"
  >
    <FeedbackRequestSummaryContent
      feedbackRequestSummary={
        Object {
          "emailWhenGrouped": true,
          "feedbackPrompt": "feedbackPrompt",
          "feedbackRequestId": 1901,
          "genre": "genre",
          "mediaType": "mediaType",
          "mediaUrl": "mediaUrl",
        }
      }
    />
  </Card>
  <EditFeedbackRequestModal
    emailWhenGrouped={true}
    feedbackPrompt="feedbackPrompt"
    feedbackRequestId={1901}
    genre="genre"
    isVisible={false}
    mediaUrl="mediaUrl"
    onCancel={[Function]}
  />
</Spin>
`;

exports[`FeedbackRequestSummary renders a summary of an existing feedback request, including media and buttons to edit and delete the request 1`] = `
<Spin
  size="default"
  spinning={false}
  wrapperClassName=""
>
  <Card
    extra={
      <RequestButtons
        mediaUrl="mediaUrl"
        onDelete={[Function]}
        onEditClick={[Function]}
      />
    }
    title="You submitted:"
  >
    <FeedbackRequestSummaryContent
      feedbackRequestSummary={
        Object {
          "emailWhenGrouped": true,
          "feedbackPrompt": "feedbackPrompt",
          "feedbackRequestId": 1901,
          "genre": "genre",
          "mediaType": "mediaType",
          "mediaUrl": "mediaUrl",
        }
      }
    />
  </Card>
  <EditFeedbackRequestModal
    emailWhenGrouped={true}
    feedbackPrompt="feedbackPrompt"
    feedbackRequestId={1901}
    genre="genre"
    isVisible={false}
    mediaUrl="mediaUrl"
    onCancel={[Function]}
  />
</Spin>
`;
